drakrun/drakrun/lib/postprocessing/capa_plugin/capa_processor.py

```diff
@@ -0,0 +1,428 @@
+        backend = capa.main.BACKEND_DOTNET
+    else:
+        backend = capa.main.BACKEND_VIV
+
+    # get the right capa feature extractor for the input: Vivisect, .NET, etc.
+    extractor = capa.loader.get_extractor(
+        dump_path,
+        input_format,
+        capa.main.OS_WINDOWS,
+        backend,
+        [],
+        False,
+        False,
+        dump_path,
+    )
+
+    # extract capabilities from the file, and ignore the returned metadata
+    capabilities, _ = capa.capabilities.common.find_capabilities(
+        rules, extractor, disable_progress=True
+    )
+
+    # return the capabilities alongside the memdump's path (for TTP categorization purposes)
+    return dump_path, capabilities
+
+
+def static_memory_dumps_capa_analysis(
+    analysis_dir: Path, rules: capa.rules.RuleSet, malware_pids: List[int] = []
+) -> Iterator[Tuple[Path, capa.render.result_document.MatchResults]]:
+    malware_dumps = list(
+        itertools.chain(
+            *(get_process_memory_dumps(analysis_dir, pid) for pid in malware_pids)
+        )
+    )
+
+    # extract all memory dumps temporarily into a dumps/ folder
+    with zipfile.ZipFile(analysis_dir / "dumps.zip", "r") as zip_ref:
+        try:
+            # try to extract the memory dumps into the analysis folder
+            zip_ref.extractall(analysis_dir)
+            dumps = analysis_dir / "dumps"
+        except PermissionError:
+            # in case of missing permissions, extract into /tmp
+            zip_ref.extractall(Path("/tmp"))
+            dumps = Path("/tmp") / "dumps"
+
+    # extract the capabilities within each memory dump, one per thread
+    pool = multiprocessing.Pool(processes=len(malware_dumps))
+    yield from pool.starmap(
+        static_capa_analysis, map(lambda dump: (dumps / dump, rules), malware_dumps)
+    )
+
+    # try to remove the temporarily created folder
+    try:
+        shutil.rmtree(dumps)
+    except PermissionError:
+        logger.debug(
+            "Permission Denied: Could not remove temporary dumps.zip extraction folder"
+        )
+
+
+def format_capa_address(address: Union[Tuple, capa.features.address.Address]) -> Dict:
+    # this method formats capa address (in the format of tuples) into a single dictionary
+    if isinstance(address, tuple):
+        return functools.reduce(
+            lambda a, b: a | b, [format_capa_address(addr) for addr in address]
+        )
+    elif isinstance(
+        address,
+        (
+            capa.features.address.AbsoluteVirtualAddress,
+            capa.features.address.RelativeVirtualAddress,
+            capa.features.address.FileOffsetAddress,
+        ),
+    ):
+        return {"address": hex(address)}
+    elif isinstance(address, capa.features.address.DNTokenAddress):
+        return {"token": address.token}
+    elif isinstance(address, capa.features.address.DNTokenOffsetAddress):
+        return {**format_capa_address(address.token), "offset": address.offset}
```

Comment:
Does DNTokenAddress have a field of type token?
https://github.com/mandiant/capa/blob/da6c6cfb4894ee40da8f53d89bbc95df6e20bb78/capa/features/address.py#L142
I think it doesn't.

Similarly, in DNTokenOffsetAddress you call format_capa_address with address.token, but it's int and doesn't match the function signature (address can't be a first parameter).

I didn't test it on a real file but - just looking at the code - I think it won't work. Maybe I'm wrong though.

Reply:
Oh right! I'm now doing:
```python
elif isinstance(address, ca.DNTokenAddress):
    return {"token": hex(address)}
elif isinstance(address, ca.DNTokenOffsetAddress):
    return {**format_capa_address(ca.DNTokenAddress(address.token)), "offset": address.offset}
```